hibernate-testing/src/main/java/org/hibernate/testing/cleaner/AbstractMySQLDatabaseCleaner.java

```diff
@@ -0,0 +1,186 @@
+	private static final String SYSTEM_SCHEMAS = "'information_schema'," +
+			"'mysql'," +
+			"'sys'," +
+			"'performance_schema'";
+	private final List<String> ignoredTables = new ArrayList<>();
+	private final Map<String, List<String>> clearingSqlsPerSchema = new HashMap<>();
+
+	@Override
+	public void addIgnoredTable(String tableName) {
+		ignoredTables.add( tableName );
+	}
+
+	@Override
+	public void clearAllSchemas(Connection connection) {
+		try {
+			clearSchema( connection, connection.getSchema() );
+		}
+		catch (SQLException e) {
+			throw new RuntimeException( e );
+		}
+	}
+
+	@Override
+	public void clearSchema(Connection connection, String schemaName) {
+		clearSchema0( connection, schemaName );
+	}
+
+	private void clearSchema0(Connection c, String schemaName) {
+		clearingSqlsPerSchema.remove( schemaName );
+		try (Statement s = c.createStatement()) {
+			// Collect schema names
+			LOG.log( Level.FINEST, "Collect table names: START" );
+			ResultSet rs = s.executeQuery(
+					"SELECT TABLE_SCHEMA, TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA = '" + schemaName + "'" );
+
+			StringBuilder sb = new StringBuilder( "DROP TABLE " );
+			if ( rs.next() ) {
+				do {
+					String tableSchema = rs.getString( 1 );
+					String tableName = rs.getString( 2 );
+					sb.append( tableSchema );
+					sb.append( '.' );
+					sb.append( tableName );
+					sb.append( ',' );
+				} while ( rs.next() );
+			}
+			else {
+				// Nothing to clear since there are no tables
+				return;
+			}
+			sb.setCharAt( sb.length() - 1, ' ' );
+			LOG.log( Level.FINEST, "Collect table names: END" );
+
+			// Disable foreign keys
+			LOG.log( Level.FINEST, "Disable foreign keys: START" );
+			s.execute( "SET FOREIGN_KEY_CHECKS = 0" );
+			LOG.log( Level.FINEST, "Disable foreign keys: END" );
+
+			LOG.log( Level.FINEST, "Dropping tables: START" );
+			String sql = sb.toString();
+			s.execute( sql );
+			LOG.log( Level.FINEST, "Dropping tables: END" );
+
+			// Enable foreign keys
+			LOG.log( Level.FINEST, "Enabling foreign keys: START" );
+			s.execute( "SET FOREIGN_KEY_CHECKS = 1" );
+			LOG.log( Level.FINEST, "Enabling foreign keys: END" );
+
+			LOG.log( Level.FINEST, "Committing: START" );
+			c.commit();
+			LOG.log( Level.FINEST, "Committing: END" );
+		}
+		catch (SQLException e) {
+			try {
+				c.rollback();
+			}
+			catch (SQLException e1) {
+				e.addSuppressed( e1 );
+			}
+
+			throw new RuntimeException( e );
+		}
+	}
+
+	@Override
+	public void clearAllData(Connection connection) {
+		clearData0( connection, null, s -> {
+			try {
+				return s.executeQuery(
+						"SELECT TABLE_SCHEMA, TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA NOT IN (" + SYSTEM_SCHEMAS + ")" );
+			}
+			catch (SQLException e) {
+				throw new RuntimeException( e );
+			}
+		} );
+	}
+
+	@Override
+	public void clearData(Connection connection, String schemaName) {
+		clearData0( connection, schemaName, s -> {
```

Comment:
Stylistically I find this difficult to read.  I know everyone will have different takes on this.  But understanding the arguments here was not instantaneous - it took me a few times looking between the call and the signature to figure it out.

Reply:
IMO, this is best:
```
clearData0(
    connection,
    schemaName,
    (tableName) -> {...}
);
```